mobile/src/main/java/net/bonysoft/doityourselfie/MainActivity.kt

```diff
@@ -1,12 +1,118 @@
 package net.bonysoft.doityourselfie
 
+
+import android.content.Intent
 import android.support.v7.app.AppCompatActivity
 import android.os.Bundle
+import android.support.v7.widget.LinearLayoutManager
+import android.view.View
+import android.widget.Toast
+import com.orhanobut.hawk.Hawk
+import kotlinx.android.synthetic.main.activity_main.*
+import kotlinx.coroutines.experimental.android.UI
+import kotlinx.coroutines.experimental.launch
+import net.bonysoft.doityourselfie.authentication.AuthenticationListener
+import net.bonysoft.doityourselfie.authentication.GoogleSignInAuthenticator
+import net.bonysoft.doityourselfie.photos.PhotosAPI
+import net.bonysoft.doityourselfie.photos.model.CompleteAlbum
+import net.bonysoft.doityourselfie.ui.AlbumAdapter
+import net.bonysoft.doityourselfie.ui.AlbumSelectedListener
+
+class MainActivity : AppCompatActivity(), AuthenticationListener, AlbumSelectedListener {
+
+    private lateinit var authenticator: GoogleSignInAuthenticator<MainActivity>
+    private lateinit var photosAPI: PhotosAPI
 
-class MainActivity : AppCompatActivity() {
+    private val adapter by lazy { AlbumAdapter(this, this) }
 
     override fun onCreate(savedInstanceState: Bundle?) {
         super.onCreate(savedInstanceState)
         setContentView(R.layout.activity_main)
+        authenticator = GoogleSignInAuthenticator.attachTo(this)
+
+        albumList.let {
+            it.adapter = adapter
+            it.layoutManager = LinearLayoutManager(this)
+        }
+
+        albumView.setListener(this)
+
+        showLoggedOutUi()
+    }
+
+    override fun showLoggedUi(token: String?) {
+        loggedInUi.visibility = View.VISIBLE
+        loggedOutUi.visibility = View.GONE
+        btnLogout.setOnClickListener {
+            authenticator.logout()
+        }
+
+        photosAPI = PhotosAPI(application, token!!, BuildConfig.DEBUG)
+        Hawk.put(TOKEN_KEY, token)
+        albumName.setText("Test Album 001")
+
+        btnCreateAlbum.setOnClickListener {
+            createAlbum(albumName.text.toString())
+        }
+
+        btnListAlbums.setOnClickListener {
+            fetchAlbums()
+        }
+    }
+
+    private fun createAlbum(albumName: String) {
+        launch(UI) {
+            waitingUi.visibility = View.VISIBLE
+            listUi.visibility = View.GONE
+            singleAlbumUi.visibility = View.GONE
+            try {
+                val response = photosAPI.createAlbum(albumName).await()
+                albumView.bindTo(response)
+            } catch (e: Exception) {
+                Toast.makeText(this@MainActivity, "ERROR: ${e.message}", Toast.LENGTH_LONG).show()
+            } finally {
+                waitingUi.visibility = View.GONE
+                singleAlbumUi.visibility = View.VISIBLE
+            }
+        }
+    }
+
+    private fun fetchAlbums() {
+        launch(UI) {
+            waitingUi.visibility = View.VISIBLE
```

Comment:
Didn't you create an extension for hide / show views?

Reply:
All this part will be deleted soon, I just need a testing environment, but it's all going away as it's not needed in the app, which should only handle logging in and passing the token to the Things app